### ISSUE TITLE
Understanding output of .pt file of YOLOv8

### ISSUE THREAD

**Issue by tdengs** [ultralytics/ultralytics#8421]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I was trying to understand the output produced by the model by running the code below

```
import torch

model_path = 'best.pt'
checkpoint = torch.load(model_path)
model = checkpoint['model'].float()  #
model.eval()

# my image size is (960, 960)
input_data = torch.randn(1,  3,  960,  960)

with torch.no_grad():
    output = model(input_data)

print(output)
```

The result I obtained indicates a tuple containing two elements. Each element within the tuple is a tensor with dimensions [1, 27, 18900]. While I understand the significance of the dimensions [1, 27, 18900], I'm unsure why there are two of these tuples.

**Comment by glenn-jocher** [comment]
@tdengs hi there! 👋

The output you're seeing is typical for YOLO models, which often provide multiple detection layers with different scales. Each tensor in the tuple corresponds to a different scale of detection, allowing the model to detect objects of various sizes more effectively.

Here's a simplified explanation of the output dimensions:
- `1` represents the batch size.
- `27` is likely the number of classes plus 4 bounding box attributes and one objectness score (e.g., for 20 classes, you'd have 20 + 4 + 1 = 25).
- `18900` is the flattened grid cells count, which comes from the height and width of the feature map times the number of anchors per cell.

The reason you have two tensors with the same dimensions is that YOLOv8 predicts at multiple scales, typically three. It seems you might be missing one scale in your output, or perhaps the model is configured differently.

If you want to inspect each detection layer separately, you can index the tuple like `output[0]` for the first scale and `output[1]` for the second scale.

I hope this clears things up! If you have more questions, feel free to ask. 😊

**Comment by tdengs** [comment]
Thank you @glenn-jocher. That clears things up for me. 

When exporting the PyTorch model to TorchScript I only get one tensor, which scale would it be? Is it the smallest one?

**Comment by glenn-jocher** [comment]
@tdengs, you're on the right track! 🌟

Indeed, the three scales you've identified in `output[1]` correspond to the multi-scale predictions YOLOv8 makes, allowing it to detect objects of various sizes. These scales are represented by the different spatial dimensions of the feature maps (`120x120`, `60x60`, and `30x30`), with `87` likely being the number of channels (`classes + 4 bbox attributes + 1 objectness score` for each anchor).

`output[0]`, on the other hand, is typically used for inference and contains the processed detections from all scales combined into a single tensor for convenience. This tensor is ready for non-maximum suppression (NMS) to filter out overlapping detections and select the most confident ones.

Your understanding is spot-on! If you have more questions or need further clarification, feel free to ask. Happy coding! 😊

**Comment by tdengs** [comment]
> @tdengs hi there! 👋
> 
> The output you're seeing is typical for YOLO models, which often provide multiple detection layers with different scales. Each tensor in the tuple corresponds to a different scale of detection, allowing the model to detect objects of various sizes more effectively.
> 
> Here's a simplified explanation of the output dimensions:
> 
> * `1` represents the batch size.
> * `27` is likely the number of classes plus 4 bounding box attributes and one objectness score (e.g., for 20 classes, you'd have 20 + 4 + 1 = 25).
> * `18900` is the flattened grid cells count, which comes from the height and width of the feature map times the number of anchors per cell.
> 
> The reason you have two tensors with the same dimensions is that YOLOv8 predicts at multiple scales, typically three. It seems you might be missing one scale in your output, or perhaps the model is configured differently.
> 
> If you want to inspect each detection layer separately, you can index the tuple like `output[0]` for the first scale and `output[1]` for the second scale.
> 
> I hope this clears things up! If you have more questions, feel free to ask. 😊

Hi @glenn-jocher, I have 23 classes, is there a reason why I'm missing a value? I think I'm missing the objections score

**Comment by glenn-jocher** [comment]
Hi @tdengs! 😊

For a model with 23 classes, the output channel should indeed be `23 (classes) + 4 (bbox attributes) + 1 (objectness score) = 28`. If you're seeing `27`, it sounds like there might be a slight discrepancy in the model configuration or the way the output is being interpreted.

Could you double-check your model's configuration to ensure it's set up for 23 classes? Sometimes, a mismatch in the expected number of classes between training and inference can lead to such issues. If everything seems correct, please share a snippet of your model configuration or the section where you define the number of classes, and we'll take a closer look! 

Here's a quick check you can do:
```python
print(model.yaml['nc'])  # This should print the number of classes your model is configured for
```

Let's get to the bottom of this!

**Comment by glenn-jocher** [comment]
Hi @tdengs! 😊

Thanks for sharing the details! From what you've described, it looks like your model and dataset are correctly set up for 23 classes. The output from the `.pt` file also suggests that the model is predicting for 23 classes, as indicated by the final convolution layers outputting 23 channels, which align with your class count.

If you're encountering unexpected behavior or outputs, it might be helpful to ensure that the inference code is correctly interpreting these outputs. Specifically, make sure that the post-processing steps (like non-maximum suppression) are correctly applied and that they're considering all 23 classes.

Here's a quick check to ensure your model's output channels align with your class count during inference:

```python
# Assuming 'model' is your loaded model and 'output' is the result of model(input_data)
output = model(input_data)
for out in output[1]:  # Loop through each scale
    print(out.shape)  # Should show something like [batch_size, 23 + 4 + 1, grid_size, grid_size]
```

This snippet helps verify that each detection layer outputs the expected number of channels. If everything looks correct but you're still facing issues, it might be worth revisiting the inference and post-processing code to ensure it's correctly handling the model's outputs.

Let me know if this helps or if you have any more questions!

**Comment by glenn-jocher** [comment]
@tdengs hi there! 😊

Sure, I'd be happy to clarify! In the context of YOLOv8, when I mentioned that the output doesn't include an explicit objectness score, I meant that the model's output tensors are designed to directly represent class probabilities, bounding box coordinates, and an overall confidence score for each detection. The "objectness" measure, which indicates the likelihood of an object being present in the bounding box, is implicitly factored into these confidence scores rather than being provided as a separate, distinct value.

In simpler terms, the confidence score you see for each detection already considers the objectness of the detection, combining it with the class probability to give you a single, comprehensive measure of confidence. This design choice streamlines the output and simplifies post-processing steps like non-maximum suppression (NMS).

Hope this clears things up! Let me know if you have any more questions. 😄

**Comment by tdengs** [comment]
Hey @glenn-jocher!

Does that mean that the output tensor having a shape of [batch size, number of classes + 4, number of detections] is correct since the objections score is implicitly factored into the class probability score?

**Comment by glenn-jocher** [comment]
@tdengs hey there! 😊

Yes, you've got it exactly right. The structure of the output tensor you've observed, `[batch size, number of classes + 4, number of detections]`, aligns perfectly with the expected output from the YOLOv8 model. Here, "number of classes + 4" accounts for both the class probabilities and the four bounding box regression values (x, y, width, height), and as you noted, the objectness score is indeed integrated within the class probabilities, streamlining the outputs.

Your training snippet and the shape of the output tensor you've shared confirm that everything is operating as intended. Your exploratory approach, confirming the tensor size through training and inference, solidifies your understanding of YOLOv8's output structure. Great job!

If you have any more questions or need further assistance, feel free to ask. Happy coding! 🚀

**Comment by tdengs** [comment]
Thank you for that @glenn-jocher. In that case, how can I calculate the mAP score without the 'objectness' score? If you could provide some guidance of how I can calculate mAP with the above tensor output it would be greatly appreciated :)

**Comment by glenn-jocher** [comment]
@tdengs hey there! 😊

Great question! With YOLOv8's output, calculating mAP (mean Average Precision) is pretty straightforward since the model outputs directly give you everything needed - class probabilities (implicitly including the objectness score) and bounding box coordinates.

For mAP calculation, you primarily focus on confidence scores and the true/false positives determined by comparing predicted boxes against ground truth using IoU (Intersection over Union) thresholds. Here's a simplified view:

1. **Determine Confidence Scores**: For each detected object, use the highest class probability as the confidence score. The model's design ensures this score already integrates the objectness concept.

2. **Match Predictions with Ground Truth**:
   - Utilize IoU to match each prediction with ground truths. 
   - A prediction is considered a true positive if IoU > IoU_threshold (commonly 0.5) and the classes match. 

3. **Calculate Precision and Recall**:
   - **Precision**: The ratio of correctly predicted positive observations to the total predicted positives.
   - **Recall**: The ratio of correctly predicted positive observations to all observations in actual class.

4. **Compute mAP**: Average the Precision across different Recall levels, usually at IoU thresholds ranging from 0.5 to 0.95 (with a step size of 0.05).

Here’s a very basic pseudo-code to get started:
```python
confidences = output[0][:, :4]  # Assuming output[0] contains [batch size, number of classes + 4, number of detections]
bbox_preds = output[0][:, 4:0]  # Bounding box predictions
# You would need your ground truth boxes and classes here.

# Calculate IoU
# Match predictions with ground truths
# Compute true positives, false positives, false negatives
# Calculate precision and recall
# Finally, compute mAP
```

Note: In practice, you’ll likely use a library or an existing tool that comes with the COCO dataset or another benchmark dataset you're using to calculate mAP more efficiently and accurately.

Hope this helps! If you have further questions, feel free to ask. Happy coding! 🚀

**Comment by ego-thales** [comment]
> Thank you @glenn-jocher. That clears things up for me!
> 
> I have done as you said and found out that output[0] is a tensor of size (1, 27, 18900) and output[1] is a list of length 3. After further inspection on each of the 3 elements of output[1]. I got the following:
> 
>     1. output[1][0].shape = torch.Size([1, 87, 120, 120])
> 
>     2. output[1][1].shape = torch.Size([1, 87, 60, 60])
> 
>     3. output[1][2].shape = torch.Size([1, 87, 30, 30])
> 
> 
> I think the 3 scales you mentioned reside in output[1], and I assume output[0] constitutes the consolidation of all 3 of the tensors? Let me know if my assumption is incorrect.

Hello, did you end up identifying the `87` from `output[1][0].size(1)`?
From a few tests with $nc\in\{8, 23, 80\}$, I figured out that it was $nc+4+60$ (the 4 corresponding to the bounding box), but I don't know what the other 60 represent.

**Comment by glenn-jocher** [comment]
Hello, yes, your assumption is correct. The three scales in `output[1]` represent different feature map resolutions, and `output[0]` consolidates these detections. The `87` in `output[1][0].size(1)` corresponds to `nc + 4 * reg_max`, where `nc` is the number of classes. For more details, refer to the linked thread.
